### PR TITLE
Adding implementation

### PR DESCRIPTION
**Pull request by nkatzman**
This PR adds the implementation for building up the unique combinations of key to values
- Asserts that the options are in the correct format and throws an error if it doesn't conform
- Uses recursion to go down the tree path of adding key values to an object
- Return case is when the currentCombination contains all keys in the options object

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  * Generate all permutations of a given environment matrix
+  * @method generateMatrix
+  * @param  {Object}       options         Object of Key => Array
+  * @return {Array}                        List of permutations of the environment
+ */
+function generateMatrix(options) {
+    const perms = [];
+    const keysToIterate = Object.keys(options);
+
+    keysToIterate.forEach((key) => {
```
Comment (stjohnjohnson) [inline index.js:47]: You could also use `joi.assert(options, joi.object().pattern(/.*/, joi.array())` but this works fine.

Comment (nkatzman) [inline index.js:47]: Ah i was wondering how to do that. I came to this implementation cause I couldn't figure that out haha

Comment (stjohnjohnson) [inline index.js:47]: Yeah, it's not really easy and clear, that's why what you did in this case is totally okay.

**Comment by stjohnjohnson** [comment]
👍 
woaah

**Comment by d2lam** [comment]
LGTM 👍  
200w-1

**Comment by minzcmu** [comment]
👍 
happy

**Pull request merged**